marlin-succinct/protocol/src/prover.rs

```diff
@@ -0,0 +1,599 @@
+
+        // extract secret of the witness
+        let mut w = witness.clone();
+        for i in 0..w.1 {w.0[i] = E::Fr::zero()}
+        let w = Evaluations::<E::Fr>::from_vec_and_domain(w.0, index.h_group).interpolate();
+
+        // prover computes za, zb polynomials
+        let mut zv = vec![vec![E::Fr::zero(); index.h_group.size()]; 2];
+
+        for i in 0..2
+        {
+            for constraint in index.compiled[i].constraints.iter()
+            {
+                zv[i][(constraint.1).0] += &(*constraint.0 * &witness.0[(constraint.1).1]);
+            }
+        }
+
+        // save public part of the witness
+        let mut public = witness.clone();
+        for i in public.1..public.0.len() {public.0[i] = E::Fr::zero()}
+
+        // prover interpolates the vectors and computes the evaluation polynomial
+        let za = Evaluations::<E::Fr>::from_vec_and_domain(zv[0].to_vec(), index.h_group).interpolate();
+        let zb = Evaluations::<E::Fr>::from_vec_and_domain(zv[1].to_vec(), index.h_group).interpolate();
+
+        // substitute ZC with ZA*ZB
+        let zv = [za.clone(), zb.clone(), &za * &zb]; 
+
+        // the transcript of the random oracle non-interactive argument
+        // the first part of the argument needs the following oracles:
+        // alpha, betta[0] in order to compute RA and evaluate Z, ZA & ZB polynomials
+
+        let mut argument = RandomOracleArgument::<E::Fr>::new(index.oracle_params.clone());
+        
+        argument.commit_scalar(&E::Fr::one()); // this commit comes from the previous proof context
+        oracles.alpha = argument.challenge();
+        // commit the public input into the argument
+        argument.commit_slice(&witness.0[0..witness.1]);
+        oracles.betta[0] = argument.challenge();
+
+        // evaluate Z, ZA & ZB polynomials
+        w_eval = w.evaluate(oracles.betta[0]);
+        za_eval = za.evaluate(oracles.betta[0]);
+        zb_eval = zb.evaluate(oracles.betta[0]);
+
+        // run the random oracle argument further to query
+        // oracles gamma[0..3]
+        argument.commit_scalar(&w_eval);
+        oracles.gamma[0] = argument.challenge();
+        argument.commit_scalar(&za_eval);
+        oracles.gamma[1] = argument.challenge();
+        argument.commit_scalar(&zb_eval);
+        oracles.gamma[2] = argument.challenge();
+
+        let mut apow = E::Fr::one();
+        let mut r: Vec<E::Fr> = (0..index.h_group.size()).map
+        (
+            |i|
+            {
+                if i > 0 {apow *= &oracles.alpha}
+                apow
+            }
+        ).collect();
+        r.reverse();
+        let ra = DensePolynomial::<E::Fr>::from_coefficients_vec(r);
+
+        // compute commitments, evaluations and openings for the first sumcheck argument
+        // proceed with the random oracle argument further to query additional oracles
+        // betta[1], batch[0]
+
+        match Self::sumcheck_1_compute (index, &ra, &zv, &z, &oracles)
```

Comment:
I believe you can similarly use the `?` operator here https://doc.rust-lang.org/edition-guide/rust-2018/error-handling-and-panics/the-question-mark-operator-for-easier-error-handling.html